### PR TITLE
Notes progress

### PR DESCRIPTION
**Pull request by sophieheb**

### CODE REVIEW

File: Gemfile.lock [hunk 1]
```diff
@@ -265,6 +266,7 @@ GEM
     thor (0.20.3)
     thread_safe (0.3.6)
     timecop (0.9.1)
+    truncate_html (0.9.3)
```
Comment (jits) [inline Gemfile.lock]: This gem seems a bit outdated and with some [potential issues](https://github.com/hgmnz/truncate_html/issues) that could affect the course output (e.g. the Unicode and emojis issues) – I notice it uses regular expressions, which makes it quite error prone (I think). Would it be worth switching to something like https://github.com/jorgemanrubia/truncato, which uses Nokogiri under the hood + is more recent and up to date?

File: app/controllers/auth_controller.rb [hunk 2]
```diff
@@ -2,7 +2,14 @@ class AuthController < ApplicationController
...
 
   def sign_in
-    session[:last_course_id] = params[:last_course_id] if params.key?(:last_course_id)
+    session.delete(:last_course_id)
+    if params.key?(:last_course_id)
+      session[:last_course_id] = params[:last_course_id]
+    elsif params.key?(:course_id)
+      session[:last_course_id] = params[:course_id]
+      session[:forwarding_url] = course_path(params[:course_id])
+    end
```
Comment (jits) [inline app/controllers/auth_controller.rb]: Nice 👍 – if you want, you can pull this out in to a separate PR, so we can get it to production quicker? (I'm assuming it's not strictly relevant/dependent on the rest of the PR).

File: app/controllers/courses_controller.rb [hunk 3]
```diff
@@ -26,5 +26,9 @@ def show
     @subscription = (current_user.subscriptions.find_by(course_slug: @course.slug) if current_user?)
 
     @vote_total = Vote.get_course_vote_total(@course.slug)
+
+    @reactions = current_user.note_reactions.where(course_slug: @course.slug).map { |r| [r.lesson_slug, r.reaction_name] }.to_h if current_user?
+
+    @progress = current_user.progresses.where(course_slug: @course.slug).map(&:lesson_slug) if current_user?
```
Comment (jits) [inline app/controllers/courses_controller.rb]: There are now (at least 3 different data objects that get initialised if there's a `current_user` present – may be worth sticking these all in a `if current_user? … end` block to contain them in one place?

File: app/models/note_reaction.rb [hunk 4]
```diff
@@ -0,0 +1,12 @@
+class NoteReaction < ApplicationRecord
+  enum reaction_name: { empowered: 0, strong: 1, knowledable: 2, magical: 3, peaceful: 4 }
```
Comment (jits) [inline app/models/note_reaction.rb]: I think it'd be good if the persisted enum values were the strings themselves, rather than integers – this way any database backups or inspecting db data directly will be more readable.

File: app/controllers/reactions_controller.rb [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+  before_action :require_authentication
+  before_action :find_or_initialize_reaction, only: %i[show update]
+
+  def show
+    update
+  end
+
+  def update
+    if @reaction.new_record?
+      @reaction[:reaction_name] = params[:reaction_name]
```
Comment (jits) [inline app/controllers/reactions_controller.rb]:
```suggestion
      @reaction.reaction_name = params[:reaction_name]
```

File: app/models/note_reaction.rb (hunk 4 above)
Comment (jits) [reply app/models/note_reaction.rb]: Also: could you add a presence validation for `reaction_name`.

File: app/controllers/reactions_controller.rb [hunk 6]
```diff
@@ -0,0 +1,29 @@
+class ReactionsController < ApplicationController
+  before_action :require_authentication
+  before_action :find_or_initialize_reaction, only: %i[show update]
+
+  def show
```
Comment (jits) [inline app/controllers/reactions_controller.rb]: I may have missed this – is this route being used anywhere?

File: app/controllers/auth_controller.rb (hunk 2 above)
Comment (sophieheb) [reply app/controllers/auth_controller.rb]: Pulled out into here: https://github.com/chaynHQ/soulmedicine/pull/162

File: Gemfile.lock (hunk 1 above)
Comment (sophieheb) [reply Gemfile.lock]: Updated :)

File: app/controllers/auth_controller.rb (hunk 2 above)
Comment (sophieheb) [reply app/controllers/auth_controller.rb]: Merged into master now.

File: app/controllers/reactions_controller.rb (hunk 6 above)
Comment (sophieheb) [reply app/controllers/reactions_controller.rb]: Yes - all links clicked in emails are get requests. I couldn't find a way around it.

Comment (jits) [reply app/controllers/reactions_controller.rb]: Ahh yes. We _may_ want to separate out the email reaction link into a separate action and add a signature / token to the link (like in the unsubscribe link), for a couple of reasons:

1. Otherwise it's possible for someone to craft a URL and have it perform the action on behalf of a user (by getting the user to click on it) – this is a form of [CSRF](https://www.owasp.org/index.php/Cross-Site_Request_Forgery_(CSRF)) although probably a minor security concern for us.
2. This `GET` link _may_ be triggered automatically by robots / search engines – of course, the way you've set this up it'll give an error (as the request won't be authenticated), but I'm not sure how that will affect SEO, etc.

Given this, do you reckon it's worth changing the way this is done? (I'm 50/50, for what it's worth).

File: app/models/note_reaction.rb (hunk 4 above)
Comment (sophieheb) [reply app/models/note_reaction.rb]: Done :)

File: app/controllers/reactions_controller.rb (hunk 6 above)
Comment (sophieheb) [reply app/controllers/reactions_controller.rb]: Hey - I found some bugs and have fixed the way this works a little bit.

Do you want to take another look and let me know if I've changed your mind at all!

File: app/views/courses/show.html.erb [hunk 7]
```diff
@@ -5,127 +5,198 @@
...
-                    </ul>
+                </div>
+                <div>
+                  <% if @subscription.present? && @subscription.lessons_delivered.include?(l.slug) %>
+                    <div class="row">
+                      <div class="col-2 pr-1">
+                        <%== emojify(':love_letter:') %>
+                      </div>
+                      <div class="col-10 pl-1">
+                        <p class="mb-0 subparagraph">You've recieved this note as an email.</p>
```
Comment (jits) [inline app/views/courses/show.html.erb]:
```suggestion
                        <p class="mb-0 subparagraph">You've received this note as an email.</p>
```

File: app/controllers/reactions_controller.rb (hunk 6 above)
Comment (jits) [reply app/controllers/reactions_controller.rb]: It feels a tad messy and overloaded. Suggestion:

- Don't reuse the `show` action for this – get rid of it, and…
- … instead, have a new action `GET set_from_email` (more explicit on what it's doing).
- … which you can now use in the email when setting a **new** reaction.
- Then in the email, don't show anything if the user already has a reaction for that lesson (my thinking here is: it's a low probability that this will happen). Or just have a link that says something like "you've already reacted to this lesson – open the lesson page to see more". This avoids having a `GET` endpoint that performs a destructive update or delete.
- In the `find_or_initialize_reaction` filter, only call `SignedGlobalID.find` if `action.to_sym == :set_from_email`, just to make sure the intention is more explicit and it only gets called then.

Also note that when attempting to find the user via the signed token we should error if no user is found, like in the Subscriptions controller: https://github.com/chaynHQ/soulmedicine/blob/88491d9ffb728a77b37b40e5abf9f7728e5b15a3/app/controllers/subscriptions_controller.rb#L63

(Give me a shout on Slack if you have any questions/thoughts on this, or if I can help in any way).

File: app/views/courses/show.html.erb [hunk 8]
```diff
@@ -5,127 +5,198 @@
...
+        <li class="breadcrumb-item"><a href="/">Home</a></li>
+        <li class="breadcrumb-item"><a href="<%= courses_path -%>">Pathways</a></li>
+        <li class="breadcrumb-item active" aria-current="page"><%= tc(@course.name) %></li>
+      </ol>
+    </nav>
+    <div class="col-4 px-0 my-auto text-right">
+      <button class="btn btn-primary dropdown-toggle" type="button" id="dropdownMenuButton" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
+        <%= LocalesService.language_name(params[:locale]) %>
+      </button>
+      <div class="dropdown-menu" aria-labelledby="dropdownMenuButton">
```
Comment (jits) [inline app/views/courses/show.html.erb]: Recommend adding the `dropdown-menu-right` class here, which forces the dropdown menu that opens to be right aligned (makes for a better layout/UI, especially on mobile).

File: app/views/courses/show.html.erb [hunk 9]
```diff
@@ -5,127 +5,198 @@
...
+              "sign in.",
+              course_subscription_path(@course),
+            )
+          %>
+        </h5>
+      <% end %>
+  </div>
 
-              <%=
-                link_to t('phrases.edit').capitalize,
```
Comment (jits) [inline app/views/courses/show.html.erb]: Can now remove the `phrases.edit` key from the translations file as I believe it's not being used anywhere else.

(I was going to mention that all strings should be pushed to the translation files, but given it's unlikely we'll ever translate the site itself anytime soon there's probably no point :))

Comment (jits) [reply app/views/courses/show.html.erb]: Oh sorry, ignore this comment – it _is_ being used in the subscription card partial.

**Comment by jits** [comment]
One UI nitipick, on mobile the following could do with more spacing (before the `Notes` title and section):

Screenshot 2019-12-16 11 05 15

**Comment by jits** [comment]
Another UI nitpick, the reactions section could do with some extra padding, both on the sides and between the options (as on mobile it could be easy to select the wrong one):

Screenshot 2019-12-16 11 08 52

**Comment by jits** [comment]
I _really_ like the new notes list (plus inline reaction / seen statuses) ❤️ 👍

File: app/views/lessons/show.html.erb [hunk 10]
```diff
@@ -1,3 +1,5 @@
+<% redirect_url = local_assigns.fetch :redirect_url, nil %>
```
Comment (jits) [inline app/views/lessons/show.html.erb]: Does this have any effect? (Since this is a top level page and not a partial, I'm guessing you can't pass it a value like this?

File: spec/factories/notereactions.rb [hunk 11]
```diff
@@ -0,0 +1,5 @@
+FactoryBot.define do
+  factory :noteReaction do
```
Comment (jits) [inline spec/factories/notereactions.rb]:
```suggestion
  factory :note_reaction do
```

… also, the suggest the filename be `note_reactions.rb`.

File: app/controllers/reactions_controller.rb (hunk 6 above)
Comment (sophieheb) [reply app/controllers/reactions_controller.rb]: Have updated again - totally agree with your thoughts. This is 100% tidied, what do you think?

File: app/views/lessons/show.html.erb (hunk 10 above)
Comment (sophieheb) [reply app/views/lessons/show.html.erb]: It wasn't necessary, good spot! Removed now.

Comment (jits) [reply app/views/lessons/show.html.erb]: Yay!

File: app/controllers/reactions_controller.rb [hunk 12]
```diff
@@ -0,0 +1,33 @@
+class ReactionsController < ApplicationController
+  before_action :require_authentication, only: %i[update destroy]
+  before_action :find_or_initialize_reaction, only: %i[set_from_email update destroy]
+
+  def set_from_email
+    @reaction.update!(reaction_name: params[:reaction_name])
```
Comment (jits) [inline app/controllers/reactions_controller.rb]: Just to check, will `.update!` work if it's a new/un-persisted record?

File: app/views/lessons/show.html.erb [hunk 13]
```diff
@@ -79,14 +79,67 @@
...
+              "create an subscription.",
+              course_subscription_path(@course),
+            )
+          %>
+        </h6>
+        <h6 class='my-2'>Or you can
+          <%=
+            link_to(
+              "sign in",
+              auth_sign_in_path(course_id: @course.to_param),
```
Comment (jits) [inline app/views/lessons/show.html.erb]: Signing in from the lesson page doesn't redirect you back to the lesson page (it takes you to the course page for that lesson).

I don't think this is show-stopping, so can always address this in a separate PR, after this gets in.

File: app/controllers/reactions_controller.rb (hunk 6 above)
Comment (jits) [reply app/controllers/reactions_controller.rb]: Looks great!

File: app/controllers/reactions_controller.rb (hunk 12 above)
Comment (sophieheb) [reply app/controllers/reactions_controller.rb]: Yeah! It does

File: app/views/lessons/show.html.erb (hunk 13 above)
Comment (sophieheb) [reply app/views/lessons/show.html.erb]: Ahh yes I did this on purpose though so they could see their previous reactions? Let's do in a separate PR

File: app/controllers/reactions_controller.rb (hunk 12 above)
Comment (jits) [reply app/controllers/reactions_controller.rb]: Ah that's good to know 👍

File: app/views/lessons/show.html.erb (hunk 13 above)
Comment (jits) [reply app/views/lessons/show.html.erb]: 👍

**Comment by jits** [comment]
Hey @sophieheb – as discussed, I've revamped + added some tests – check out https://github.com/chaynHQ/soulmedicine/pull/146/commits/33ac2c43284c2cae786db1c58f73439ee6bd321d when you have a sec.

Also, I noticed that my local `git commit` command was complaining because I didn't have the `erb_lint` gem installed (from when you added the check in https://github.com/chaynHQ/soulmedicine/pull/144) – really useful to have this! At first I thought I may as well add it to the Gemfile (in the `:development, :test` group), but then I noticed it also installed other gems like `better_html` which I don't think we should add to the codebase as it could impact template rendering, etc. So it probably makes sense to **not** have it in the Gemfile, but we should probably add to the README as a prerequisite – what do you think?
